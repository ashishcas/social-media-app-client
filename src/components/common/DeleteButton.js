import React, {useState, useEffect} from 'react';

import { Button, Icon, Confirm, Popup } from 'semantic-ui-react';

import { gql, useMutation } from '@apollo/client';

import { Link } from 'react-router-dom';

import { FETCH_POSTS } from '../utils/graphqlQueries';


const DELETE_POST_MUTATION = gql`
  mutation deletePost($postId: ID!) {
    deletePost(postId: $postId)
  }
`
const DELETE_COMMENT_MUTATION=gql`
  mutation deleteComment($postId: ID!, $commentId: ID!) {
    deleteComment(postId: $postId, commentId: $commentId) {
      id
      comments {
        id
        username
        createdAt
        body
      }
      commentCount
    }
  }
`;



export const DeleteButton = (props) =>{

    const [ confirmOpen, setConfirmOpen] = useState(false);

    const mutation = props.commentId ? DELETE_COMMENT_MUTATION: DELETE_POST_MUTATION;
    const [deletePost] = useMutation(mutation, {
        update(proxy){       
            setConfirmOpen(false);
            if(!props.commentId){
                let data = proxy.readQuery({
                    query: FETCH_POSTS
                })
                let getPostsData = data.getPosts.filter(p => p.id !== props.postId);
                proxy.writeQuery({query: FETCH_POSTS, data:{
                    getPosts: [...getPostsData],
                }});
            }
            if(props.callback){
                props.callback();
            }
        },
        variables: { postId: props.postId , commentId: props.commentId}
    });


    return(
        <>
        <Popup 
        content={ props.commentId ? 'Delete Comment': 'Delete post'} 
        inverted
        trigger={            
          <Button as='div' 
            color="grey" 
            floated="right"
            onClick={() => { setConfirmOpen(true)}}
            >
              <Icon name="trash" style={{margin: 0}} />
            </Button>
            }
          />
            <Confirm 
                open={confirmOpen}
                onCancel = { () => { setConfirmOpen(false)}}
                onConfirm = { deletePost }
            />
        </>
    )
}