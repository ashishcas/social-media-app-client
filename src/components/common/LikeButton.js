import React, {useState, useEffect} from 'react';

import { Button, Icon, Label, Popup } from 'semantic-ui-react';

import { gql, useMutation } from '@apollo/client';

import { Link } from 'react-router-dom';


const LIKE_POST_MUTATION = gql`
  mutation likePost($postId: ID!) {
    likePost(postId: $postId) {
      id
      likes {
        id
        username
      }
      likeCount
    }
  }
`


export const LikeButton = ({ user, post: { id , likeCount, likes }}) =>{

    const [ liked, setLiked] = useState(false);

    useEffect(() =>{
        if(user && likes.find(like => like.username === user.username)){
            setLiked(true)
        }else {
            setLiked(false);
        }
    },[user, likes])


    const [likePost] = useMutation(LIKE_POST_MUTATION, {
        variables: { postId: id }
    });

    const likeButton = user ? (
      liked ?
      (  <Button color='red' >
            <Icon name='heart' />
        </Button>
      ) : (
        <Button color='red' basic>
            <Icon name='heart' />
        </Button>
      )
    ): (
        <Button as={Link} to='/login' color='red' basic>
            <Icon name='heart' />
        </Button>
    )
    return(
        <>
        <Popup 
        content={ liked ? 'Unlike': 'like'} 
        inverted
        trigger={            
          <Button as='div' labelPosition='right' onClick={likePost}>
          {likeButton}
          <Label as='a' basic color='red' pointing='left'>
              {likeCount}
          </Label>
          </Button>
            }
          />

        </>
    )
}