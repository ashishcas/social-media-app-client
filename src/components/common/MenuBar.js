import React, { useState, useContext } from 'react';
import { Menu } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import { AuthContext } from '../context/auth';
import { isMobile } from 'react-device-detect';


function MenuBar(){
    const { user , logout } = useContext(AuthContext);
    const pathname = window.location.pathname;
    const path = pathname === '/' ? 'home' : pathname.substr(1);
    const [activeItem, setActiveItem] = useState(path);
    const  handleItemClick = (e, { name }) => { setActiveItem(name)};
    const menuBar  = user ? (
      <div>
      <Menu stackable pointing secondary size="massive" color="teal" >
        <Menu.Item
          name={user?.username}
          active
          as={Link}
          to="/"
        />
        <Menu.Menu position='right'>
          <Menu.Item
              name='Messages'
              as={Link}
              to="/messages"
          />
        </Menu.Menu>
        <Menu.Menu position='right'>
          <Menu.Item
              name='logout'
              onClick={logout}
          />
        </Menu.Menu>
        
      </Menu>
    </div>
    ) : (
      <>
        <div>
        <Menu  stackable pointing secondary size="massive" color="teal">
          <Menu.Item
            name='home'
            active={activeItem === 'home'}
            onClick={handleItemClick}
            as={Link}
            to="/"
          />
          <Menu.Menu position='right'>
            <Menu.Item
                name='login'
                active={activeItem === 'login'}
                onClick={handleItemClick}
                as={Link}
                to="/login"
            />
            <Menu.Item
              name='register'
              active={activeItem === 'register'}
              onClick={handleItemClick}
              as={Link}
              to="/register"
            />
          </Menu.Menu>
        </Menu>
      </div>
    </>
    )

    return menuBar;
}

export default MenuBar;