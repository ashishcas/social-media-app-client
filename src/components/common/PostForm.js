import React, {useState} from 'react';

import { Form, Button } from 'semantic-ui-react';

import { gql, useMutation } from '@apollo/client';

import { FETCH_POSTS } from '../utils/graphqlQueries';


const CREATE_POST_MUTATION = gql`
    mutation createPost($body: String!){
        createPost(body: $body){
            id 
            body
            createdAt
            username 
            likes{
                id
                username
                createdAt
            }
            likeCount
            comments {
                id
                body
                username
                createdAt
            }
            commentCount
        }
    }

`

const PostForm = () => {
    const [fields , setFields] = useState({
        body: '',
    });

    const [ errors , setErrors ] = useState('');

    const handleChange =(event) => {
        setFields({...fields,[event.target.name]: event.target.value})
    };



    const [ createPost, { error  }] = useMutation(CREATE_POST_MUTATION,{
        update(proxy, result){
            const data = proxy.readQuery({
                query: FETCH_POSTS
            })
            // data.getPosts = [result.data.createPost, ...data.getPosts];
            proxy.writeQuery({query: FETCH_POSTS, data:{
                 getPosts: [result.data.createPost, ...data.getPosts],
            }});
            fields.body= ''
        },
        onError(err){
            setErrors(err?.graphQLErrors[0]?.message)
        },
        variables:fields
    });

    const onSubmit=(event) => {
        event.preventDefault();
        createPost();
    }

    return(
        <>
            <Form onSubmit={onSubmit}>
                    <h2>Create a Post</h2>
                    <Form.Field>
                            <Form.Input 
                                placeholder="Hi world"
                                name="body"
                                onChange={handleChange}
                                value={fields.body}
                                error={errors ? true : false}
                                />
                            <Button type="submit" color="teal">
                                Submit
                            </Button>
                    </Form.Field>
            </Form>
            {
                errors && (
                    <div className="ui error message" style={{marginBottom: 20}}>
                        <ul className="list">
                            <li>{errors}</li>
                        </ul>
                    </div>
                )
            }
        </>
    )
}

export default PostForm;