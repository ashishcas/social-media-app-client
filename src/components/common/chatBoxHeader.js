// the header compone
import {Button, Card , Icon , Label, Image, Popup } from 'semantic-ui-react';


const ChatBoxHeader = ({profileIcon,nameOfFriend}) => {
    return (
        <Card>
            <Card.Content>
                <Image floated='left' size='mini' src={profileIcon || 'https://react.semantic-ui.com/images/avatar/large/molly.png'} alt="profile"/>
                <div className="chat-box-header-name">{nameOfFriend}</div>
            </Card.Content>
        </Card>
    )
}

export default ChatBoxHeader;