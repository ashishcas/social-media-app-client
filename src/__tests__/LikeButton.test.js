import React from 'react';
import { shallow, mount, render} from 'enzyme';
import { LikeButton } from '../components/common/LikeButton';
import { MockedProvider, LIKE_POST_MUTATION } from '@apollo/client/testing';


describe('snapshot testing like button',()=>{

    const mocks = []
    const mockMutation = {
        request: {
            query: LIKE_POST_MUTATION,
            variables: { postId: '1' },
          },
          result: {
            data: { post: { id: 1, name: 'Buck', breed: 'poodle' } },
          },
    };
    const post = {id:'1', likeCount: 0, likes:[]};

    it('should render like button', () => {

        const ShallowWrapper = shallow(
            <MockedProvider mocks={mocks}>
            <LikeButton user="somename" post={post}/>
            </MockedProvider>
        );
        expect(ShallowWrapper).toMatchSnapshot();
    });

    it('test onclick after render like button', () => {

        const ShallowWrapper = render(
            <MockedProvider mocks={mocks}>
            <LikeButton user="somename" post={post}/>
            </MockedProvider>
        );
        console.log((ShallowWrapper.find('#likeButton').length));
        // ;
        // expect(ShallowWrapper).toMatchSnapshot();
        // ShallowWrapper.unmount();
    });
});
