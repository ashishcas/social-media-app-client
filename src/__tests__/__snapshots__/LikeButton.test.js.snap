// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `ShallowWrapper {}`;

exports[`snapshot testing like button should render like button 1`] = `ShallowWrapper {}`;

exports[`snapshot testing like button test onclick after render like button 1`] = `
LoadedCheerio {
  "0": Node {
    "attribs": Object {
      "class": "ui right labeled button",
      "role": "button",
      "tabindex": "0",
    },
    "children": Array [
      Node {
        "attribs": Object {
          "class": "ui red basic button",
        },
        "children": Array [
          Node {
            "attribs": Object {
              "aria-hidden": "true",
              "class": "heart icon",
            },
            "children": Array [],
            "name": "i",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "aria-hidden": undefined,
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "aria-hidden": undefined,
              "class": undefined,
            },
          },
        ],
        "name": "button",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Node {
          "attribs": Object {
            "class": "ui red left pointing basic label",
          },
          "children": Array [
            Node {
              "data": "0",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "a",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Node {
        "attribs": Object {
          "class": "ui red left pointing basic label",
        },
        "children": Array [
          Node {
            "data": "0",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "a",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Node {
          "attribs": Object {
            "class": "ui red basic button",
          },
          "children": Array [
            Node {
              "attribs": Object {
                "aria-hidden": "true",
                "class": "heart icon",
              },
              "children": Array [],
              "name": "i",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "aria-hidden": undefined,
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "aria-hidden": undefined,
                "class": undefined,
              },
            },
          ],
          "name": "button",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": Node {
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
    },
    "prev": null,
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
      "role": undefined,
      "tabindex": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
      "role": undefined,
      "tabindex": undefined,
    },
  },
  "_root": LoadedCheerio {
    "0": Node {
      "children": Array [
        Node {
          "attribs": Object {},
          "children": Array [
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "head",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "body",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
            Node {
              "attribs": Object {},
              "children": Array [],
              "name": "body",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": Node {
                "attribs": Object {},
                "children": Array [],
                "name": "head",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {},
                "x-attribsPrefix": Object {},
              },
              "type": "tag",
              "x-attribsNamespace": Object {},
              "x-attribsPrefix": Object {},
            },
          ],
          "name": "html",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {},
          "x-attribsPrefix": Object {},
        },
      ],
      "name": "root",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
